### PR TITLE
feat: added webhooks events

### PR DESCRIPTION
**Pull request by sagarnaikjuspay**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Checklist

- [x] I ran `npm run re:build`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: src/screens/HyperSwitch/APIUtils/APIUtils.res [hunk 1]
```diff
@@ -144,6 +144,15 @@ let getURL = (
       }
     | _ => ""
     }
+  | CONNECTOR_EVENT_LOGS =>
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/APIUtils/APIUtils.res:147]:
```
| CONNECTOR_EVENT_LOGS =>  switch id {
      | Some(payment_id) => `analytics/v1/outgoing_webhook_event_logs?payment_id=${payment_id}`
      | None => ""
      }
```

Why to add a methodType check we can remove when needed we can add.

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res [hunk 2]
```diff
@@ -1,8 +1,18 @@
-open OrderUtils
-open APIUtils
-open LogicUtils
 @module("js-sha256") external sha256: string => string = "sha256"
+
+open PaymentLogsTypes
+let getLogType = dict => {
+  if dict->Dict.get("request_id")->Belt.Option.isSome {
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:5]: `Belt.Option.isSome` use directly `Option.isSome`

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res [hunk 3]
```diff
@@ -1,8 +1,18 @@
...
+    PAYMENTS
+  } else if dict->Dict.get("component")->Belt.Option.isSome {
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:7]: `Belt.Option.isSome` use directly `Option.isSome`

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res [hunk 4]
```diff
@@ -34,27 +43,29 @@ module PrettyPrintJson = {
...
-            <pre
-              className={`${overrideBackgroundColor} p-3 text-jp-gray-900 dark:bg-jp-gray-950 dark:bg-opacity-100 ${isTextVisible
-                  ? "overflow-visible "
-                  : `overflow-clip  h-fit ${maxHeightClass}`} text-fs-13 text font-medium`}>
-              {parsedJson->React.string}
-            </pre>
+            {
+              let style = isTextVisible
+                ? "overflow-visible "
+                : `overflow-clip  h-fit ${maxHeightClass}`
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:63]:
```
 let style = isTextVisible
                ? "overflow-visible "
                : `overflow-clip  h-fit ${maxHeightClass}`
                
```
Can we write this code outside JSX. Then {} will not be needed

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogsUtils.res [hunk 5]
```diff
@@ -10,3 +10,26 @@ let nameToURLMapper = (urlName, ~payment_id, ~merchant_id="", ()) => {
...
+  "platform",
+  "version",
+]
+
+let sortByCreatedAt = (log1: Js.Json.t, log2: Js.Json.t) => {
+  open LogicUtils
+  let getKey = dict => dict->getDictFromJsonObject->getString("created_at", "")->Js.Date.fromString
+  let keyA = log1->getKey
+  let keyB = log2->getKey
+  if keyA < keyB {
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogsUtils.res:28]: `LogicUtils.compareLogic() `

can be called directly need to pass KeyA and keyB

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res [hunk 6]
```diff
@@ -323,139 +362,165 @@ let make = (~paymentId, ~createdAt) => {
...
-                jsonToDisplay=responseObject
-                headerText={Some(currentSelectedType === Payment ? "Response body" : "Metadata")}
-              />
+            <UIUtils.RenderIf condition={!(logDetails.response->isEmptyString)}>
+              {
+                let headerText = switch selectedOption.optionType {
+                | PAYMENTS => "Response body"
+                | WEBHOOKS => "Request body"
+                | SDK => "Metadata"
+                }->Some
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:521]: Can this be put outside JSX?

`  let headerText = switch selectedOption.optionType {
                | PAYMENTS => "Response body"
                | WEBHOOKS => "Request body"
                | SDK => "Metadata"
                }->Some`
                
And then {} will be removed

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res [hunk 7]
```diff
@@ -323,139 +362,165 @@ let make = (~paymentId, ~createdAt) => {
...
-    screenState customUI={<NoDataFound message="No logs available for this payment" />}>
-    {if paymentId->HSwitchOrderUtils.isTestPayment {
+    screenState customUI={<NoDataFound message="No logs available for this PAYMENTS" />}>
+    {if paymentId->HSwitchOrderUtils.isTestPayment || logs.contents->Array.length === 0 {
       <div
         className="flex items-center gap-2 bg-white w-full border-2 p-3 !opacity-100 rounded-lg text-md font-medium">
         <Icon name="info-circle-unfilled" size=16 />
         <div className={`text-lg font-medium opacity-50`}>
-          {"No logs available for this payment"->React.string}
+          {"No logs available for this PAYMENTS"->React.string}
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:474]: `{"No logs available for this PAYMENTS"->React.string}`


It should be Payment only i guess. Only one payment we are dealing at a time I guess

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res [hunk 8]
```diff
@@ -323,139 +362,165 @@ let make = (~paymentId, ~createdAt) => {
...
   React.useEffect0(() => {
     getDetails()->ignore
     None
   })
 
+  open OrderUtils
   <PageLoaderWrapper
-    screenState customUI={<NoDataFound message="No logs available for this payment" />}>
-    {if paymentId->HSwitchOrderUtils.isTestPayment {
+    screenState customUI={<NoDataFound message="No logs available for this PAYMENTS" />}>
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:468]: {"No logs available for this PAYMENTS"->React.string}

It should be Payment only i guess. Only one payment we are dealing at a time I guess

**Review by PritishBudhiraja — CHANGES_REQUESTED** [review]

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res [hunk 9]
```diff
@@ -78,44 +89,36 @@ module PrettyPrintJson = {
...
-    let logType = paymentDetailsValue->Dict.get("request_id")->Belt.Option.isSome ? Payment : Sdk
+    let logType = paymentDetailsValue->getLogType
     let apiName = switch logType {
-    | Payment => paymentDetailsValue->getString("api_flow", "default value")->camelCaseToTitle
-    | Sdk => paymentDetailsValue->getString("event_name", "default value")
-    }->PaymentUtils.nameToURLMapper(~payment_id=paymentId, ())
+    | PAYMENTS => paymentDetailsValue->getString("api_flow", "default value")->camelCaseToTitle
+    | SDK => paymentDetailsValue->getString("event_name", "default value")
+    | WEBHOOKS => paymentDetailsValue->getString("outgoing_webhook_event_type", "default value")
+    }->PaymentLogsUtils.nameToURLMapper(~payment_id=paymentId, ())
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:111]: PaymentLogsUtils it's already open

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res [hunk 10]
```diff
@@ -125,48 +128,66 @@ module ApiDetailsComponent = {
       ->Dict.fromArray
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:128]: LogicUtils.getJsonFromArrayOfJson

```
->Dict.fromArray
      ->Js.Json.object_
```

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res [hunk 11]
```diff
@@ -204,65 +229,79 @@ module ApiDetailsComponent = {
...
 
-  let sourceMapper = source => {
-    switch source {
-    | "ORCA-LOADER" => "HYPERLOADER"
-    | "ORCA-PAYMENT-PAGE"
-    | "STRIPE_PAYMENT_SHEET" => "PAYMENT_SHEET"
-    | other => other
+  let fetchWebhooksLogsData = async _ => {
+    try {
+      let webhooksLogsUrl = APIUtils.getURL(
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:274]: Already open APIUtils

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res [hunk 12]
```diff
@@ -204,65 +229,79 @@ module ApiDetailsComponent = {
...
+  let (selectedOption, setSelectedOption) = React.useState(_ => {
+    value: "",
+    optionType: PAYMENTS,
+  })
+  let (screenState, setScreenState) = React.useState(_ => PageLoaderWrapper.Loading)
 
   let fetchPaymentLogsData = async _ => {
     try {
-      setScreenState1(_ => PageLoaderWrapper.Loading)
       let paymentLogsUrl = APIUtils.getURL(
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:255]: Already open APIUtils

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res [hunk 13]
```diff
@@ -323,139 +362,165 @@ let make = (~paymentId, ~createdAt) => {
...
-        fetchSdkLogsData()->ignore
-      } else {
-        setScreenState1(_ => PageLoaderWrapper.Success)
-        setScreenState2(_ => PageLoaderWrapper.Success)
+    if !(paymentId->HSwitchOrderUtils.isTestPayment) {
+      let screenState = switch (
+        await fetchPaymentLogsData(),
+        await fetchSdkLogsData(),
+        await fetchWebhooksLogsData(),
+      ) {
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:387]: Can you define variable of these and then write switch case for that - 

`  await fetchPaymentLogsData(),
        await fetchSdkLogsData(),
        await fetchWebhooksLogsData(),`
        
Better for code readability and understanding

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res [hunk 14]
```diff
@@ -323,139 +362,165 @@ let make = (~paymentId, ~createdAt) => {
...
+        | SDK => {
+            let request =
+              initialData
+              ->Dict.toArray
+              ->Array.filter(entry => {
+                let (key, _) = entry
+                filteredKeys->Array.includes(key)->not
+              })
+              ->Dict.fromArray
+              ->Js.Json.object_
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:426]: `->Dict.fromArray
              ->Js.Json.object_`
              
LoUtils.getJsonFromArrayOfJson

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res [hunk 15]
```diff
@@ -323,139 +362,165 @@ let make = (~paymentId, ~createdAt) => {
...
+                logsDataLength={logs.contents->Array.length - 1}
               />
             })
             ->React.array}
           </div>
         </div>
         <UIUtils.RenderIf
-          condition={responseObject->String.length > 0 || requestObject->String.length > 0}>
+          condition={logDetails.response->String.length > 0 ||
+            logDetails.request->String.length > 0}>
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:503]: ` condition={!(logDetails.response->isEmptyString) ||
            !(logDetails.request->isEmptyString)}>`

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res [hunk 16]
```diff
@@ -323,139 +362,165 @@ let make = (~paymentId, ~createdAt) => {
...
+            <UIUtils.RenderIf
+              condition={!(logDetails.request->isEmptyString) &&
+              selectedOption.optionType !== WEBHOOKS}>
               <PrettyPrintJson
-                jsonToDisplay=requestObject
-                headerText={Some(currentSelectedType === Payment ? "Request body" : "Event")}
-                maxHeightClass={responseObject->String.length > 0 ? "max-h-25-rem" : ""}
+                jsonToDisplay=logDetails.request
+                headerText={Some(selectedOption.optionType === PAYMENTS ? "Request body" : "Event")}
+                maxHeightClass={logDetails.response->String.length > 0 ? "max-h-25-rem" : ""}
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:512]: `maxHeightClass={!(logDetails.response->isEmptyString) ? "max-h-25-rem" : ""}`

**Review by PritishBudhiraja — CHANGES_REQUESTED** [review]

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res (hunk 13 above)
Comment (sagarnaikjuspay) [reply src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:415]: not needed

**Review by PritishBudhiraja — DISMISSED** [review]

File: src/utils/LogicUtils.res [hunk 17]
```diff
@@ -492,6 +492,8 @@ let isEmptyDict = dict => {
 
 let isEmptyString = str => str->String.length === 0
 
+let isNonEmptyString = str => str->String.length > 0
```
Comment (PritishBudhiraja) [inline src/utils/LogicUtils.res:495]: This is not req operations can be performed via doing ~not of the above. Just a recommendation

Comment (sagarnaikjuspay) [reply src/utils/LogicUtils.res:495]: its better to have for better readability

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res [hunk 18]
```diff
@@ -323,87 +399,174 @@ let make = (~paymentId, ~createdAt) => {
...
+      })
+      logs.contents = logs.contents->Array.concat(logsArr)
 
-      setScreenState2(_ => PageLoaderWrapper.Success)
+      PageLoaderWrapper.Success
     } catch {
     | Js.Exn.Error(e) =>
-      let err = Js.Exn.message(e)->Option.getWithDefault("Failed to Fetch!")
-      setScreenState2(_ => PageLoaderWrapper.Error(err))
+      let err = Js.Exn.message(e)->Belt.Option.getWithDefault("Failed to Fetch!")
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:413]: Use option directly

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res [hunk 19]
```diff
@@ -78,197 +90,261 @@ module PrettyPrintJson = {
...
         ~id=Some(paymentId),
         (),
       )
+      let connectorLogsArray = (await fetchDetails(connectorLogsUrl))->getArrayFromJson([])
+      logs.contents = logs.contents->Array.concat(connectorLogsArray)
+
+      PageLoaderWrapper.Success
+    } catch {
+    | Js.Exn.Error(e) =>
+      let err = Js.Exn.message(e)->Belt.Option.getWithDefault("Failed to Fetch!")
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:331]: Use option directly

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res [hunk 20]
```diff
@@ -78,197 +90,261 @@ module PrettyPrintJson = {
...
-
-      setScreenState1(_ => PageLoaderWrapper.Success)
+      logs.contents = logs.contents->Array.concat(paymentLogsArray)
+
+      PageLoaderWrapper.Success
     } catch {
     | Js.Exn.Error(e) =>
-      let err = Js.Exn.message(e)->Option.getWithDefault("Failed to Fetch!")
-      setScreenState1(_ => PageLoaderWrapper.Error(err))
+      let err = Js.Exn.message(e)->Belt.Option.getWithDefault("Failed to Fetch!")
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:290]: Option directly

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res [hunk 21]
```diff
@@ -78,197 +90,261 @@ module PrettyPrintJson = {
...
+              </p>
+            </div>
             {switch logType {
-            | Sdk => <p className=headerStyle> {apiName->React.string} </p>
-            | Payment =>
-              <p className=headerStyle>
-                <span className="font-bold mr-2"> {method->String.toUpperCase->React.string} </span>
+            | SDK =>
+              <p className={`${headerStyle} mt-1 ${isSelected ? "" : "opacity-80"}`}>
+                {apiName->LogicUtils.camelCaseToTitle->React.string}
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:240]: Already open LogicUtils

**Review by PritishBudhiraja — CHANGES_REQUESTED** [review]
Please make a diff PR for the new dependency addition

File: public/hyperswitch/assets/copy.svg
[the remark this answers is not in the document]
Comment (sagarnaikjuspay) [reply public/hyperswitch/assets/copy.svg:1]: this is a diff icon

File: src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res [hunk 22]
```diff
@@ -34,11 +45,11 @@ module PrettyPrintJson = {
 
     let copyParsedJson =
       <div onClick={_ => handleOnClickCopy(~parsedValue=parsedJson)} className="cursor-pointer">
-        <img src={`/assets/CopyToClipboard.svg`} />
+        <img src={`/assets/copy.svg`} className="w-9 h-5" />
```
Comment (PritishBudhiraja) [inline src/screens/HyperSwitch/PaymentLogs/PaymentLogs.res:48]: Can we add this icon in solid.svg and use from that place?

**Review by PritishBudhiraja — APPROVED** [review]

**Review by jainlokesh318 — APPROVED** [review]

**Review by JeevaRamu0104 — APPROVED** [review]

**Pull request merged**